Code/GraphMol/MolInterchange/Parser.cpp

```diff
@@ -0,0 +1,469 @@
+      else if (propVal.value.IsString())
+        mol->setProp(propVal.name.GetString(), propVal.value.GetString());
+    }
+  }
+
+  if (molval.HasMember("extensions")) {
+    for (const auto &propVal : molval["extensions"].GetArray()) {
+      if (!propVal.HasMember("name"))
+        throw FileParseException(
+            "Bad Format: representation has no name member");
+      if (propVal["name"].GetString() == std::string("rdkitRepresentation")) {
+        readRDKitRepresentation(mol, propVal, params);
+      }
+      if (propVal["name"].GetString() == std::string("partialCharges")) {
+        readPartialCharges(mol, propVal, params);
+      }
+    }
+  }
+  mol->updatePropertyCache(false);
+  mol->setProp(common_properties::_StereochemDone, 1);
+}
+
+std::vector<boost::shared_ptr<ROMol>> DocToMols(
+    rj::Document &doc, const JSONParseParameters &params) {
+  std::vector<boost::shared_ptr<ROMol>> res;
+
+  // some error checking
+  if (!doc.IsObject())
+    throw FileParseException("Bad Format: JSON should be an object");
+  if (!doc.HasMember("commonchem"))
+    throw FileParseException("Bad Format: missing header in JSON");
+  if (!doc["commonchem"].HasMember("version"))
+    throw FileParseException("Bad Format: missing version in JSON");
+  if (doc["commonchem"]["version"].GetInt() != currentMolJSONVersion)
+    throw FileParseException("Bad Format: bad version in JSON");
+
+  rj::Value atomDefaults_;
+  if (rj::GetValueByPointer(doc, "/defaults/atom")) {
+    atomDefaults_ = *rj::GetValueByPointer(doc, "/defaults/atom");
+    if (!atomDefaults_.IsObject())
+      throw FileParseException("Bad Format: atomDefaults is not an object");
+  }
+  const DefaultValueCache atomDefaults(atomDefaults_);
+
+  rj::Value bondDefaults_;
+  if (rj::GetValueByPointer(doc, "/defaults/bond")) {
+    bondDefaults_ = *rj::GetValueByPointer(doc, "/defaults/bond");
+    if (!bondDefaults_.IsObject())
+      throw FileParseException("Bad Format: bondDefaults is not an object");
+  }
+  const DefaultValueCache bondDefaults(bondDefaults_);
+
+  if (doc.HasMember("molecules")) {
+    if (!doc["molecules"].IsArray())
+      throw FileParseException("Bad Format: molecules is not an array");
+    for (const auto &molval : doc["molecules"].GetArray()) {
+      RWMol *mol = new RWMol();
+      processMol(mol, molval, atomDefaults, bondDefaults, params);
+      mol->updatePropertyCache(params.strictValenceCheck);
+      mol->setProp(common_properties::_StereochemDone, 1);
+      res.push_back(boost::shared_ptr<ROMol>(static_cast<ROMol *>(mol)));
+    }
+  }
+
+  return res;
+}
+
+}  // end of anonymous namespace
+
+std::vector<boost::shared_ptr<ROMol>> JSONDataStreamToMols(
+    std::istream *inStream, const JSONParseParameters &params) {
+  PRECONDITION(inStream, "no stream");
+
+  rj::IStreamWrapper isw(*inStream);
+  rj::Document doc;
+  doc.ParseStream(isw);
+
+  return DocToMols(doc, params);
+}
+std::vector<boost::shared_ptr<ROMol>> JSONDataToMols(
```

Comment:
Might make sense to start thinking about how to use unique_ptr for these types of interfaces.

The downside is that we can't return a `std::vector<unique_ptr<>>` directly.